core/src/avm2/regexp.rs

```diff
@@ -204,17 +208,68 @@ impl<'gc> RegExp<'gc> {
         ret
     }
 
-    pub fn replace(
+    /// Implements string.replace(regex, replacement) where the replacement is
+    /// a function.
+    pub fn replace_fn(
+        &mut self,
+        activation: &mut Activation<'_, 'gc, '_>,
+        text: AvmString<'gc>,
+        f: &FunctionObject<'gc>,
+    ) -> Result<AvmString<'gc>, Error> {
+        self.replace_with_fn(activation, &text, |activation, txt, m| {
+            let args = std::iter::once(Some(&m.range))
+                .chain((m.captures.iter()).map(|x| x.as_ref()))
+                .map(|o| match o {
+                    Some(r) => {
+                        AvmString::new(activation.context.gc_context, &txt[r.start..r.end]).into()
+                    }
+                    None => "".into(),
+                })
+                .chain(std::iter::once(m.range.start.into()))
+                .chain(std::iter::once((*txt).into()))
+                .collect::<Vec<_>>();
+            let r = f.call(activation.global_scope(), &args, activation)?;
+            return Ok(WString::from(
+                r.coerce_to_string(activation).unwrap_or_default().as_wstr(),
```

Comment:
I think we'd want to propagate exceptions here as well.

Reply:
Done.